File: 2.两数之和(链表)/golang/2.go
package main

//  Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head = new(ListNode)
	crt := head
	var inherit int
	for {
		// 退出条件：l1，l2都用完了，且进位用完了
		if l1 == nil && l2 == nil && inherit == 0 {
			break
		}
		tmpVal := inherit
		if l1 != nil {
			tmpVal = tmpVal + l1.Val
			// l1 当前节点已使用完成，进入下个节点
			l1 = l1.Next
		}
		if l2 != nil {
			tmpVal = tmpVal + l2.Val
			// l2 当前节点已使用完成，进入下个节点
			l2 = l2.Next
		}
		inherit = tmpVal / 10
		var node = new(ListNode)
		node.Val = tmpVal % 10
		crt.Next = node
		crt = node
	}
	return head.Next
}

func main() {
}

File: 2.两数之和(链表)/c/2.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    struct ListNode* head = malloc(sizeof(struct ListNode));
    head->next = NULL;
    struct ListNode* crt = head;
    int inherit = 0;
    while(true) {
        if (l1 == NULL && l2 == NULL && inherit == 0) {
            break;
        }
        int tmpVal = inherit;
        if (l1 != NULL) {
            tmpVal = tmpVal + l1->val;
            l1 = l1->next;
        }
        if (l2 != NULL) {
            tmpVal = tmpVal + l2->val;
            l2 = l2->next;
        }
        inherit = tmpVal / 10;
        struct ListNode* node = malloc(sizeof(struct ListNode));
        node->val = tmpVal % 10;
        node->next = NULL;
        crt->next = node;
        crt = node;
    }
    return head->next;
}
